src/swappy_allocator.rs

```diff
@@ -0,0 +1,1195 @@
+            // Ordering: `maybe_swapped` must be set before the page is created.
+            std::sync::atomic::fence(Ordering::Release);
+            pages.push(
+                Page::new(
+                    &self.0.path,
+                    self.0.page_seq.fetch_add(1, Ordering::Relaxed),
+                    std::cmp::max(DEFAULT_PAGE_SIZE, layout.size()),
+                )
+                .ok_or(AllocError)?,
+            );
+        }
+        debug_assert!(self.0.maybe_swapped.load(Ordering::Relaxed));
+        match pages.last_mut().unwrap().allocate(layout) {
+            None => {
+                pages.push(
+                    Page::new(
+                        &self.0.path,
+                        self.0.page_seq.fetch_add(1, Ordering::Relaxed),
+                        std::cmp::max(DEFAULT_PAGE_SIZE, layout.size()),
+                    )
+                    .ok_or(AllocError)?,
+                );
+                pages.last_mut().unwrap().allocate(layout).ok_or(AllocError)
+            }
+            Some(r) => Ok(r),
+        }
+    }
+}
+
+impl SwappyAllocator<Global> {
+    pub fn new(path: &Path, budget: usize) -> SwappyAllocator<Global> {
+        Self::new_over(path, budget, Global)
+    }
+}
+
+unsafe impl<A: Allocator> Allocator for SwappyAllocator<A> {
+    #[inline]
+    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        // Always use mem_allocator to allocate empty pointer.
+        if layout.size() > 0
+            && self.0.mem_usage.fetch_add(layout.size(), Ordering::Relaxed) + layout.size()
+                > self.0.budget
+        {
+            if layout.size() == 0 {
+                Ok(NonNull::slice_from_raw_parts(layout.dangling(), 0))
+            } else {
+                match self.allocate_swapped(layout) {
+                    r @ Ok(_) => {
+                        self.0.mem_usage.fetch_sub(layout.size(), Ordering::Relaxed);
+                        r
+                    }
+                    _ => self.0.mem_allocator.allocate(layout),
+                }
+            }
+        } else {
+            self.0.mem_allocator.allocate(layout)
+        }
+    }
+
+    #[inline]
+    unsafe fn deallocate(&self, ptr: NonNull<u8>, layout: Layout) {
+        if self.is_swapped(ptr, false) {
+            let mut pages = self.0.pages.lock();
+            // Find the page it belongs to, then deallocate itself.
+            for i in 0..pages.len() {
+                if pages[i].contains(ptr) {
+                    if pages[i].deallocate(ptr) {
+                        let page = pages.remove(i);
+                        page.release(&self.0.path);
+                    }
+                    if pages.is_empty() {
+                        self.0.maybe_swapped.store(false, Ordering::Relaxed);
+                    }
+                    return;
+                }
+            }
+        }
+        self.0.mem_usage.fetch_sub(layout.size(), Ordering::Relaxed);
```

Comment:
address space should be equal to `usize::MAX`, so I think overflow is impossible.

Reply:
I'm just afraid of bug that somewhere forget or fail(due to race condition considering memory order) to add the `mem_usage`, then it will underflow

Reply:
Assume there is a bug, the damage is not that much (only some performance overhead). But if we use wrapping sub here, it might be difficult to figure out what went wrong. So I think we should let it fail loudly.